_posts/2021-08-07-build-a-protected-web-app-with-reactjs-and-forgerock.md

```diff
@@ -0,0 +1,462 @@
+
+```js
+<form
+  className="cstm_form"
+  onSubmit={(event) => {
+    event.preventDefault();
+    const nextStep = FRAuth.next(step);
+    setStep(nextStep);
+  }}
+>
+```
+
+You may be asking, “How did the user’s input values get added to the `step` object?” Let’s take a look at the component for rendering the username input. Open up the `Text` component: `components/journey/text.js`. Notice how special methods are being used on the callback object. These are provided as convenience methods for getting and setting data.
+
+```js
+export default function Text({ callback, inputName }) {
+  const [state] = useContext(AppContext);
+  const existingValue = callback.getInputValue();
+
+  function setValue(event) {
+    callback.setInputValue(event.target.value);
+  }
+
+  return (
+    <div className={`cstm_form-floating form-floating mb-3`}>
+      <input
+        className={`cstm_form-control form-control ${validationClass} bg-transparent ${state.theme.textClass} ${state.theme.borderClass}`}
+        defaultValue={existingValue}
+        id={inputName}
+        name={inputName}
+        onChange={setValue}
+        placeholder={textInputLabel}
+      />
+      <label htmlFor={inputName}>{textInputLabel}</label>
+    </div>
+  );
+}
+```
+
+The two important items to focus on is the `callback.getInputValue()` and the `callback.setInputValue()`. The `getInputValue` is used to display any existing value that may be provided by ForgeRock, and the `setInputValue` is used to set the user’s input on the callback while they are typing (`onChange`). The reason this works, even though we are not communicating this value to the `Form` component, is due to us passing the `callback` to the components by “reference” (not by “value”). So, any mutation of the callback object within the `Text` (or `Password`) component is also contained within the `step` object in `Form`.
+
+NOTE: You may be thinking, “That’s not very idiomatic React! Shared, mutable state is bad.” And, yes, you are correct, but we are taking advantage of this to keep everything simple (and this guide from being too long), so I hope you can excuse the pattern.
+
+Each type of callback type has its own collection of methods for getting and setting data in addition to a base set methods that the SDK automatically adds to response from the ForgeRock server. For more information about these callback methods, [see our API documentation](https://sdks.forgerock.com/javascript/api-reference-core/index.html), or [the source code in Github](https://github.com/ForgeRock/forgerock-javascript-sdk/tree/master/src/fr-auth/callbacks), for more details.
+
+That that the form is rendering and submitting, we need to add an additional condition to our `Form` component. This condition will handle the success result of the authentication journey.
+
+```js
+if (!step) {
+  // Loading ...
+} else if (step.type === 'LoginSuccess') {
+  return <Alert message="Success! You're logged in." type='success' />;
```

Comment:
Keep single quote consistent and I think we need to wrap in the braces for props
```suggestion
  return <Alert message={'Success! You're logged in.'} type={'success'} />;
```